VSRAD.Package/Server/BreakState.cs

```diff
@@ -10,14 +11,62 @@ public sealed class BreakState
         public string StatusString { get; }
         public int ExitCode { get; }
         public DateTime ExecutedAt { get; } = DateTime.Now;
+        public uint WaveSize { get; }
+        public uint DimX { get; }
+        public uint DimY { get; }
+        public uint DimZ { get; }
+        public uint GroupSize { get; } // TODO: handle 3d group sizes
+        public bool NDRange3D { get; }
 
-        public BreakState(BreakStateData breakStateData, long totalElapsedMilliseconds, string statusString)
+        private static readonly Regex StatusFileRegex = new Regex(@"grid size \((?<gd_x>\d+), (?<gd_y>\d+), (?<gd_z>\d+)\)\s+group size \((?<gp_x>\d+), (?<gp_y>\d+), (?<gp_z>\d+)\)\s+wave size (?<wv>\d+)\s+comment (?<comment>.+)", RegexOptions.Compiled);
+
+        private BreakState(BreakStateData breakStateData, long totalElapsedMilliseconds, string statusFileContents)
         {
             Data = breakStateData;
             TotalElapsedMilliseconds = totalElapsedMilliseconds;
             ExecElapsedMilliseconds = 0;
-            StatusString = statusString;
             ExitCode = 0;
+
+            var match = StatusFileRegex.Match(statusFileContents);
+
+            if (match.Success)
+            {
+                var gridX = uint.Parse(match.Groups["gd_x"].Value);
+                var gridY = uint.Parse(match.Groups["gd_y"].Value);
+                var gridZ = uint.Parse(match.Groups["gd_z"].Value);
+                var groupX = uint.Parse(match.Groups["gp_x"].Value);
+                var groupY = uint.Parse(match.Groups["gp_y"].Value);
+                var groupZ = uint.Parse(match.Groups["gp_z"].Value);
+                WaveSize = uint.Parse(match.Groups["wv"].Value);
+                NDRange3D = gridY != 0 && gridZ != 0;
+
+                if ((NDRange3D && (groupY == 0 || groupZ == 0 || groupY > gridY || groupZ > gridZ))
+                    || groupX > gridX || WaveSize > groupX || groupX == 0 || WaveSize == 0 || gridX == 0)
+                    throw new ArgumentException();
+
+                GroupSize = groupX;
+                DimX = gridX / groupX;
+                DimY = NDRange3D ? gridY / groupY : 0;
+                DimZ = NDRange3D ? gridZ / groupZ : 0;
+                StatusString = match.Groups["comment"].Value;
+            }
+            else if (!string.IsNullOrEmpty(statusFileContents))
+            {
+                throw new ArgumentException();
+            }
+        }
+
+        public static BreakState GetBreakState(BreakStateData breakStateData, long totalElapsedMilliseconds, string statusFileContents)
```

Comment:
It's better to return `Result<BreakState>` and show an error message from the calling code (which handles UI). You can make the constructor accept all fields and do the validation here, so you won't even need to use exceptions, just `return new Error("...")`.

Reply:
Thanks for the suggestion, implemented in 7c0b3f5